process/block/preprocess/rewardsHandler.go

```diff
@@ -345,23 +345,50 @@ func (rtxh *rewardsHandler) createRewardFromFees() []data.TransactionHandler {
 
 // createProtocolRewards creates the protocol reward transactions
 func (rtxh *rewardsHandler) createProtocolRewards() []data.TransactionHandler {
-	consensusRewardAddresses := rtxh.address.ConsensusRewardAddresses()
+	consensusRewardData := rtxh.address.ConsensusShardRewardData()
 
 	consensusRewardTxs := make([]data.TransactionHandler, 0)
-	for _, address := range consensusRewardAddresses {
+	for _, address := range consensusRewardData.Addresses {
 		rTx := &rewardTx.RewardTx{}
 		rTx.Value = rewardValue
 		rTx.RcvAddr = []byte(address)
 		rTx.ShardId = rtxh.shardCoordinator.SelfId()
-		rTx.Epoch = rtxh.address.Epoch()
-		rTx.Round = rtxh.address.Round()
+		rTx.Epoch = consensusRewardData.Epoch
+		rTx.Round = consensusRewardData.Round
 
 		consensusRewardTxs = append(consensusRewardTxs, rTx)
 	}
 
-	rtxh.mutGenRewardTxs.Lock()
-	rtxh.protocolRewards = consensusRewardTxs
-	rtxh.mutGenRewardTxs.Unlock()
+	return consensusRewardTxs
+}
+
+// createProtocolRewardsForMeta creates the protocol reward transactions
+func (rtxh *rewardsHandler) createProtocolRewardsForMeta() []data.TransactionHandler {
+	metaRewardsData := rtxh.address.ConsensusMetaRewardData()
+	consensusRewardTxs := make([]data.TransactionHandler, 0)
+
+	for _, metaConsensusSet := range metaRewardsData {
+		for _, address := range metaConsensusSet.Addresses {
+			shardId, err := rtxh.address.ShardIdForAddress([]byte(address))
+			if err != nil {
+				log.Error(err.Error())
+				continue
```

Comment:
if this is an error, should not it break ?

Reply:
Normally this should not happen when the staking contract is available, because the address must be valid to register with the contract.

While we use config files, if an invalid address was added as a staking address, or the address is missing, then the shard ID for that address cannot be calculated and this translates in not being able to create a reward transaction for that address. 

The question is if whenever such a node is part of the consensus group if we should cancel the respective round, or we can just ignore the wrong address.